server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceType.java

```diff
@@ -85,6 +87,15 @@
     ValuesSource replaceMissing(ValuesSource valuesSource, Object rawMissing, DocValueFormat docValueFormat,
                                 LongSupplier nowSupplier);
 
+    /**
+     * Attempts to return a reader function for the indexed data of this field.  Some aggregations are able to use this as an optimization
+     * instead of relying on doc values, when the index sort order matches that of the aggregation.
+     *
+     * @param fieldType The field type we want to get a reader for
+     * @return null if we can't get a reader (e.g. because the field is the wrong type), otherwise a point reader function.
+     */
+    default Function<byte[], Number> getPointReader(MappedFieldType fieldType) { return null; }
```

Comment:
I'm not convinced this is the right place to add extensibility here.  The other option I'm considering, we could make `parsePoint` a method on `MappedFieldType` (or, maybe a method that returns a reference to `parsePoint` or null, depending on if it's implemented for that field).  That would let us get rid of the instance type checking, and provide a way for plugins that add fields which use CoreVSTs to add custom point readers.

Reply:
I agree that it should be the `MappedFieldType`'s responsibility. It is built by the `FieldMapper` which is ultimately the only thing that actually knows how the index was built.